internal/controller/clusterinstance_controller.go

```diff
@@ -239,6 +240,38 @@ func (r *ClusterInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 
 	// Only update the ObservedGeneration when all the above processes have been successfully executed
 	if clusterInstance.Status.ObservedGeneration != clusterInstance.ObjectMeta.Generation {
+
+		// Capture and compare snapshot of ClusterInstance.Spec
+		lastSpecJSON, exists := clusterInstance.Annotations[v1alpha1.LastClusterInstanceSpecAnnotation]
+		lastAppliedSpec := &v1alpha1.ClusterInstanceSpec{}
+		if exists {
+			if err := json.Unmarshal([]byte(lastSpecJSON), lastAppliedSpec); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to unmarshal last-applied Spec: %w", err)
+			}
+		}
+
+		if !exists || !equality.Semantic.DeepEqual(clusterInstance.Spec, lastAppliedSpec) {
+			currentSpecJSON, err := json.Marshal(clusterInstance.Spec)
+			if err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to marshal current ClusterInstance spec: %w", err)
+			}
+
+			log.Sugar().Infof("Updating  %s annotation to %v", v1alpha1.LastClusterInstanceSpecAnnotation,
+				string(currentSpecJSON))
+			patch := client.MergeFrom(clusterInstance.DeepCopy())
+			clusterInstance.Annotations[v1alpha1.LastClusterInstanceSpecAnnotation] = string(currentSpecJSON)
+			if err := r.Client.Patch(ctx, clusterInstance, patch); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to update %s annotation: %w",
+					v1alpha1.LastClusterInstanceSpecAnnotation, err)
+			}
+
+			// Re-fetch updated ClusterInstance
+			if err := r.Get(ctx, req.NamespacedName, clusterInstance); err != nil {
+				log.Error("Failed to get ClusterInstance", zap.Error(err))
+				return requeueWithError(err)
+			}
+		}
```

Comment:
This all feels like it should be in its own function since nothing else really depends on anything in here.
Plus that would trim this `Reconcile` method down a bit which is getting quite large

Reply:
fair enough - will factor this out in a separate function!